src/controllers/EvidenceUploadController.ts

```diff
@@ -34,32 +39,98 @@ export class EvidenceUploadController extends BaseController<OtherReason> {
         return appeal.reasons?.other;
     }
 
+    private async renderUploadError(appeal: Appeal, text: string): Promise<void> {
+        const that = this;
+        const validationResult: ValidationResult = new ValidationResult([
+            new ValidationError('file', text)
+        ]);
+
+        return await that.renderWithStatus(UNPROCESSABLE_ENTITY)(
+            that.template, {
+                ...this.prepareViewModelFromAppeal(appeal),
+                ...this.httpContext.request.body,
+                validationResult
+            }
+        );
+    }
+
     protected getExtraActionHandlers(): Record<string, FormActionHandler | FormActionHandlerConstructor> {
         const that = this;
+
+        const fileTooLargeError: string = 'File size must be smaller than 4MB';
+        const fileNotSupportedError: string = 'The selected file must be a TXT, DOC, PDF, JPEG or PNG';
+        const tooManyFilesError: string = 'You can only select up to 10 files at the same time';
+
         return {
             'upload-file': {
                 async handle(request: Request, response: Response): Promise<void> {
-                    await parseFormData(request, response);
-
-                    if (!request.file) {
-                        response.redirect(request.route.path);
-                        return;
-                    }
 
-                    const id = await that.fileTransferService.upload(request.file.buffer, request.file.originalname);
+                    const maxNumberOfFiles: number = Number(getEnvOrThrow('MAX_NUMBER_OF_FILES'));
 
                     const appeal: Appeal = request.session
                         .chain(_ => _.getExtraData())
                         .map<ApplicationData>(data => data[APPLICATION_DATA_KEY])
                         .map(data => data.appeal)
                         .unsafeCoerce();
 
+                    try {
+                        await parseFormData(request, response)
+                    } catch (error) {
+                        if (error.message === 'File not supported') {
+                            return await that
+                                .renderUploadError(appeal, fileNotSupportedError);
+                        } else if (error.message === 'File too large') {
+                            return await that
+                                .renderUploadError(appeal, fileTooLargeError);
+                        }
+                    }
+
+                    if (!request.file) {
+                        response.redirect(MOVED_TEMPORARILY, request.route.path);
```

Comment:
Shouldn't we render validation error now when we have validation? It is quite surprising for page to reload. Please get in touch with Bart or UX about this. Prototype does not have validation message cause it uses JS version where upload is triggered on file selection automatically but here we should have some validation not a reload IMO.

Reply:
Will do